nmtwizard/preprocess/prepoperator.py

```diff
@@ -22,105 +22,147 @@ def get_operator_params(config):
-                if operator and operator.is_applied():
+                operator = self._build_operator(op)
+                if operator:
                     self._ops.append(operator)
                 if preprocess_exit_step and i == preprocess_exit_step:
                     break
 
-        if process_type=="postprocess":
-            self._ops = reversed(self._ops)
-
-            self.start_state, self._build_state = self._build_state, self.start_state
-
-            # Add pure postprocessing operators.
-            postprocess_config = config.get("postprocess")
-            if postprocess_config:
-                for op in postprocess_config:
-                    op = op.copy()
-                    # Explicitely mark as a postprocess operator.
-                    op['postprocess'] = True
-                    operator = self._build_operator(op, process_type)
-                    if operator and operator.is_applied():
-                        self._ops.append(operator)
-
 
-    def _build_operator(self, operator_config, process_type):
+    def _build_operator(self, operator_config):
         op = get_operator_type(operator_config)
         params = get_operator_params(operator_config)
         operator = None
         if op == "length_filter":
-            operator = LengthFilter(params, process_type)
+            operator = LengthFilter(params)
         elif op == "tokenization":
-            operator =  Tokenizer(params, process_type, self._build_state)
+            operator =  Tokenizer(params, self._build_state)
         # TODO : all other operators
         else:
             # TODO : warning or error ?
             logger.warning('Unknown operator \'%s\' will be ignored.' % op)
         return operator
 
+
     def __call__(self, tu_batch):
-        for op in self._ops:
-            tu_batch = op(tu_batch)
+        for i, op in enumerate(self._ops):
+            tu_batch = op(tu_batch, self._process_type)
         return tu_batch
 
 
+class InferencePipeline(Pipeline):
+
+    def __init__(self, config, process_type=ProcessType.INFERENCE):
+
+        self._process_type = process_type
+
+        # TODO : do we really need those now ?
+        self.start_state = { "src_tok_config" : None,
+                             "tgt_tok_config" : None }
+
+        # Current state of pipeline.
+        # Passed to and modified by operator initializers if necessary.
+        self._build_state = dict(self.start_state)
+
+        self._build_preprocess_pipeline(config)
+
+
+
+class PostprocessPipeline(InferencePipeline):
+
+    def __init__(self, config):
+
+        super(PostprocessPipeline, self).__init__(config, ProcessType.POSTPROCESS)
+
+        # Reverse preprocessing operators.
+        self._ops = reversed(self._ops)
+
+        # TODO
+        self.start_state, self._build_state = self._build_state, self.start_state
+        # Add pure postprocessing operators.
+        postprocess_config = config.get("postprocess")
```

Comment:
There should be a function that takes a list of op config and return a list of ops. The logic is currently duplicated with the preprocess.